### DISCUSSION TITLE
mobile app support?

### DISCUSSION THREAD

**Discussion by shimon-elbaz** [mediacms-io/mediacms#975] [Show and tell]
hi, i am looking for a solution to use this with a mobile app (android mainly, but can be ios as well)
does anyone know of an app which can be used?

i dont want to use the browser at the moment.

**Comment by OfficialMikeJ** [comment]
I second this too. I would like to know if there are any plans in terms of an app that will be integrated with MediaCMS. would be nice for Android to get one first then the AppStore later on. I would mainly use it for Android myself. This is something I have been wanting to know for some time now.

**Comment by calinbule** [comment]
Hey there. Is there any news on this?

**Comment by mrjoshuap** [comment]
This is an interesting idea, though I wonder if there are benefits over just using a mobile browser?  Are there specific things needed that are limited to a mobile "app"?

My experience is mostly mobile and it far exceeds my use case, so I'm curious what features would make it compelling on a mobile outside of the web.

My thoughts lean towards perhaps easier download management for offline viewing, but again, I can do that already via the browser.

What are you thinking?

**Comment by mrjoshuap** [comment]
Ok... so I'm sort of growing into this.  I stubbed out a mobile app, have video playback working (still troubleshooting audio) and even support connections to multiple MediaCMS instances.  I opted to use Flutter so it would theoretically be able to build support for iOS and Android (among others).

A couple things:

- I can implement MUCH faster caching and "infinite scrolling" that loads all media faster than the web app can load a couple "infinite scrolls," especially with large media libraries.
- I think it might be easier (and less server overhead) to offload some of the AI capabilities I want to the client devices.
- I can do cool things with the video player (like swipe to skip/rewind)

Here's a screen recording of the app running in a iOS simulator.

https://github.com/user-attachments/assets/b5d80e09-3414-43a1-8c2c-62a52a48e02b

**Reply by mgogoulos** [reply]
can you upload the recording somewhere else? it's not playing here...

**Reply by mrjoshuap** [reply]
Try this:

http://mrjoshuap.com/mediacms/mediacms-app-smaller.mp4

**Reply by mrjoshuap** [reply]
If it's possible, I'd love to have a test user account to test with on demo.mediacms.io :-D

I do have auth working on my local instance, but I was planning on setting the default "instance" to the demo site (was going to disable logins for it though).  On that, is there an API call that can determine if registration is possible?  For now, I have it just send to the registration page.  If registration is disabled on the site, I'd just as soon throw an error on the client.

**Comment by mrjoshuap** [comment]
Alrighty!  Talk about fun.  This was actually pretty straightforward.  I've got a prototype running on both iOS and Android.

![Screenshot 2026-01-02 at 18 00 13](https://github.com/user-attachments/assets/442fd610-e9a9-47c8-9b46-b43ca81d3a19)
![Screenshot 2026-01-02 at 18 00 31](https://github.com/user-attachments/assets/bb61e6bd-84cc-4337-9b40-b73ef08347e2)
![Screenshot 2026-01-02 at 18 00 40](https://github.com/user-attachments/assets/b73cb94a-0ab5-49c7-9869-2b9f3f32b4aa)
![Screenshot 2026-01-02 at 18 01 15](https://github.com/user-attachments/assets/729a6f25-80f0-4847-bfd5-27bc6ae644b2)
![Screenshot 2026-01-02 at 18 01 23](https://github.com/user-attachments/assets/67478e76-a9f2-46f4-aef2-20d771228039)

**Comment by mrjoshuap** [comment]
Here's a longer video of my latest prototype.  It's about 4 minutes long and slightly too big to upload here.   I still have a couple bugs to resolve but core functionality works.

https://mrjoshuap.com/mediacms/mediacms-app-prototype-demo.mp4

Honestly, I kind of like it.  I have noticed a few problems with the API though... it doesn't seem to implement all of the features I need, search doesn't seem implemented, I can't seem to grab content for a category/tag, I'm not sure I can "Report" media via API, and the middleware is not proxy aware (see related proxy/auth/api issues).  Might be others, but it's late.

For my purposes, my mediacms instance has over 15k videos and the website gets REALLY slow while browsing media lists (think My Media).  I do NOT have that problem with the mobile app, it easily handles it and stays VERY responsive.

I'm open to some thoughts or feedback.

----

# MediaCMS Mobile App - Prototype Release Overview

## High-Level Overview

The MediaCMS Mobile App is a mobile client that brings the power of MediaCMS to iOS and Android devices. This prototype release demonstrates the core functionality of a native mobile experience for MediaCMS instances, offering media browsing, playback, offline downloads, and multi-instance management.  It also features configurable caching of media previews, gifs and metadata to help reduce the strain on the MediaCMS instances.

### Platform Support

- **iOS**: 12.0+ (native AVPlayer integration)
- **Android**: API 21+ (ExoPlayer integration)
- **Features**: Responsive layouts for phones and tablets

---

## Feature Comparison: Mobile App vs Website

| Feature | Mobile App | Website | Notes |
|---------|-----------|---------|-------|
| **Authentication** | ✅ | ✅ | Login, register, anonymous access, token management |
| **Multi-Instance Support** | ✅ | ❌ | Connect to multiple MediaCMS instances |
| **Media Browsing** | ✅ | ✅ | Home screen with featured, categories, recent media |
| **Media Search** | ✅ | ✅ | Category-based search with filters |
| **Media Detail View** | ✅ | ✅ | Full media information, tags, categories |
| **Video Playback** | ✅ | ✅ | HLS streaming with adaptive quality |
| **Audio Playback** | ✅ | ✅ | Audio player with artwork and background support |
| **Image Viewing** | ✅ | ✅ | Image viewer with pinch-to-zoom |
| **PDF Viewing** | ✅ | ✅ | PDF viewer with page navigation |
| **Playback Controls** | ✅ | ✅ | Play/pause, seek, volume, speed (0.5x-2x) |
| **Subtitle Support** | ✅ | ✅ | VTT subtitle display and selection |
| **Chapter Navigation** | ✅ | ✅ | Chapter markers and navigation |
| **Quality Selection** | ✅ | ✅ | HLS adaptive quality selection |
| **Resume Playback** | ✅ | ✅ | Resume from last position |
| **Full-Screen Mode** | ✅ | ✅ | Full-screen player with orientation handling |
| **Background Audio** | ⚠️ (Planned) | ❌ | Audio service integration (pending) |
| **Picture-in-Picture** | ❌ | ❌ | Not yet implemented |
| **Offline Downloads** | ✅ | ❌ | Download media for offline viewing (mobile-only) |
| **Download Management** | ✅ | ❌ | Download queue, progress tracking, storage management |
| **Playlists** | ✅ | ✅ | Create, view, and manage playlists |
| **Comments** | ❌ | ✅ | Comment viewing and posting (planned) |
| **Media Upload** | ⚠️ (Placeholder) | ✅ | Upload screen exists but not fully implemented |
| **User Profiles** | ✅ | ✅ | View and edit user profiles |
| **Profile Editing** | ✅ | ✅ | Edit name, bio, and profile information |
| **My Media** | ✅ | ✅ | View and manage user's own media |
| **Categories & Tags** | ✅ | ✅ | Browse by categories and tags |
| **Settings** | ✅ | ✅ | App configuration and preferences |
| **Data Management** | ✅ | ✅ | Clear cache, purge data, storage management |
| **Logging System** | ✅ | ❌ | Comprehensive logging with viewer (app-specific for debugging) |
| **Instance Switching** | ✅ | ❌ | Quick instance switching |
| **Responsive Layout** | ✅ | ✅ | Adaptive layouts for tablets and phones |
| **Pull-to-Refresh** | ✅ | ✅ | Refresh media feeds |
| **Infinite Scroll** | ✅ | ✅ | Paginated media loading |
| **Media Actions** | ✅ | ✅ | Like, dislike, report media |
| **Share Media** | ✅ | ✅ | Share media links |
| **AI Features** | ❌ (Planned) | ❌ | On-device AI features (planned) |
| **Chromecast** | ❌ | ❌ | Casting support (planned) |

**Legend:**
- ✅ = Fully Implemented
- ⚠️ = Partially Implemented / In Progress
- ❌ = Not Available

---

## Key Mobile-Specific Features

### 1. **Offline Downloads**
- Download media for offline viewing
- Background download support
- Download queue management
- Progress tracking and resume capability
- Storage management and cleanup

### 2. **Multi-Instance Management**
- Connect to multiple MediaCMS instances
- Quick instance switching
- Instance-specific authentication
- Preserved route state when switching

### 3. **Native Media Playback**
- Hardware-accelerated video rendering
- Background audio playback (pending notification controls)
- Native player controls integration
- Platform-optimized performance

### 4. **Mobile-Optimized UI**
- Responsive layouts (single column on phones, grid on tablets)
- Touch-optimized controls
- Gesture support (pinch-to-zoom, swipe navigation)
- Mobile-first navigation patterns

### 5. **Enhanced User Experience**
- Animated preview GIFs with configurable delay
- Shimmer loading states
- Configurable media cache timeouts
- Resume playback prompts with customizable delays
- Comprehensive logging system for debugging

---

## Implementation Status Summary

### ✅ Completed Core Features
- Authentication and user management
- Media browsing and discovery
- Full-featured media player (video, audio, image, PDF)
- Profile management
- Settings and preferences
- Offline downloads
- Playlist management
- Search functionality
- Multi-instance support

### ⚠️ In Progress / Partial
- Background audio with notification controls
- Media upload functionality
- Multi-instance data persistence improvements

### ❌ Planned / Not Started
- Comments system
- AI-powered features (on-device transcription, translation)
- Picture-in-Picture mode
- Chromecast support
- Advanced upload features (chunked upload, resume)

---

## Technical Highlights

### State Management
- Reactive state with Riverpod
- Code generation for type safety
- Provider-based dependency injection
- Efficient cache invalidation strategies

### Data Layer
- Repository pattern for API abstraction
- Local database with Drift (SQLite)
- Secure credential storage
- Offline-first architecture for downloads

### Media Playback
- Global player singleton for seamless playback
- Adaptive quality streaming (HLS)
- Subtitle and chapter support
- Playback history and resume functionality

### User Interface
- Material Design 3
- Dark/light theme support
- Responsive layouts
- Smooth animations and transitions
- Accessibility considerations

---

## Known Limitations (Prototype)

1. **Comments**: Not yet implemented (UI structure exists)
2. **Upload**: Placeholder screen only, full upload functionality pending
3. **Background Audio**: Audio service integrated but notification controls pending
4. **Picture-in-Picture**: Not implemented
5. **Chromecast**: Not implemented

---

## Development Notes

This is a **prototype release** demonstrating the core capabilities of a native MediaCMS mobile client. The foundation is solid with a clean architecture, type-safe state management, and comprehensive media playback support. Future releases would focus on completing pending features and enhancing the user experience.

**Comment by mrjoshuap** [comment]
I'll just leave this here...

https://mediacms.app

**Reply by D1reW0lf** [reply]
Hi. 
I noticed that https://mediacms.app no longer exists. 
Is your mobile still being developed?

**Comment by Yiannistaos** [comment]
Αmazing work, @mrjoshuap! 👏

I was wondering about your tech stack choice, do you plan to stick with Flutter/Dart for the long run, or have you considered a React Native version as well? 

Also, for anyone looking to contribute, would you say deep knowledge of Dart is essential, or are you planning to keep the architecture simple enough for others to jump in? 

Do you have any plans to release the source code on GitHub?

**Reply by mrjoshuap** [reply]
Full disclosure: I'm not super familiar with mobile development in general.

I went with Flutter/Dart because it seemed to have all of the necessary bits to play the media (primarily HLS streaming out of the box), access the API and easily support iOS and Android...  I didn't really "consider" much else.  I honestly wouldn't have even put React on the list because I find it slow and too resource intensive. (I assume that's what the MediaCMS frontend is written in?). Due to PTSD from web development 20 years ago during the peak of the "browser wars," I tend to avoid all JavaScript-like languages...

Right now, I do not have much experience to determine what a simple or complex architecture looks like!  I'd love to keep it simple even if just for me!

Regarding open sourcing it, I'm still undecided and open to discussion.  Most of the development and deployment requires paid subscriptions to the appropriate app stores (Apple + Google).  That said, where would it go?  I'm calling it an "unofficial" mobile app because we haven't really talked about it.

My initial release was going to be closed, primarily because I'm bouncing back and forth with the App Store review processes.  I've never built a mobile app to deploy on an App Store (most of what I've done is just an app I need for my own purposes).  I do have bias towards open source and I didn't want to really tackle that until I had something to show :-D

I also wanted to get all of the features I wanted implemented.

There is also a number of API changes/updates/adds I'd like to see incorporated into MediaCMS itself.  Some of the things like returning media for a specific category/tag are unclear to me.  I want to do more sorting and filtering (duration, categories, tags, etc).

My main motivation was to be able to offload some of my AI needs onto the client devices (transcription, object detection/tagging, recommendations, and "title/description cleanup").

What are your thoughts?

**Comment by mgogoulos** [comment]
@mrjoshuap I'd like to ask more on the performance issues you mention - when visiting the website. Is it because it takes time to load the media (images), or you suspect something DB related?

**Reply by mrjoshuap** [reply]
@mgogoulos  I don't think it DB related.  When I browse large media lists, such as "My Media" (where I have thousands of videos), the first four or five times it loads for the "infinite list", it's fine.  At some point though it starts taking a long time to load the next set and the browser (depending on which one) either pauses without any responses (brave) or I get shown a "Wait for the page to be responsive or close" (safari) dialog.  It'll eventually start responding and be responsive, up until it has to load the next set of media, where it does it again.

I just sort of chalked it up to "JavaScript, you are one fickle mistress and have terrible memory management"... and to be honest, was one of the reasons I opted to try the mobile app route -- well that and I REALLY want to add more sorting and filtering options...

I've not really done a ton of investigation into it, but while I was toying around the with mobile app, I opted for a "go ahead and download and cache all media metadata" option initially (rather than the current "infinite scroll and cache all media metadata").  While the full load took a few seconds to load the media (API is paginated), all of them were VERY responsive.  I then turned on the infinite scroll in the mobile app (where it uses the pagination instead of loading the whole list) and can flick down through them all as fast as I can go and it never exhibits the same behavior.

I can't show it on my running instance, however, I do have a branch (see the admin dashboard PR #1457) with a "media generator" that I can use to generate a huge number of "fake" media to see if I can replicate.

**Comment by MRunkehl** [comment]
Hi, just wanted to ask how the Android app is going? I'm strongly interested and still waiting.
